packages/language-server/lib/configuration.js

```diff
@@ -0,0 +1,61 @@
+/**
+ * @typedef {import('unified').Pluggable} Pluggable
+ * @typedef {import('unified').PluggableList} PluggableList
+ */
+
+import {loadPlugin} from 'load-plugin'
+
+/**
+ * Load remark plugins from a configuration object.
+ *
+ * @param {string} cwd
+ *   The current working directory to resolve plugins from.
+ * @param {unknown} config
+ *   The object that defines the configuration.
+ * @returns {Promise<PluggableList | undefined>}
+ *   A list of unified plugins to use.
+ */
+export async function loadPlugins(cwd, config) {
+  if (typeof config !== 'object') {
+    return
+  }
+
+  if (!config) {
+    return
+  }
+
+  if (!('plugins' in config)) {
+    return
+  }
+
+  let pluginConfig = config.plugins
+
+  if (typeof pluginConfig !== 'object') {
+    return
+  }
+
+  if (!pluginConfig) {
+    return
+  }
+
+  const pluginArray = Array.isArray(pluginConfig)
+    ? pluginConfig
+    : Object.entries(pluginConfig)
+
+  /** @type {Promise<Pluggable>[]} */
+  const plugins = []
+  for (const maybeTuple of pluginArray) {
+    const [name, ...options] = /** @type {unknown[]} */ ([]).concat(maybeTuple)
+
+    if (typeof name !== 'string') {
+      continue
+    }
+
+    plugins.push(
+      loadPlugin(name, {prefix: 'remark', cwd}).then(
+        (plugin) => /** @type {Pluggable} */ ([plugin, ...options])
+      )
+    )
+  }
+  return Promise.all(plugins)
```

Comment:
This looks a bit naïve, e.g., what about presets? Where are these loaded from? Do people need them in each project?

The code used by the engine is quite complex (https://github.com/unifiedjs/unified-engine/blob/5468193d0e000d321718e5612c10362b8bd7df3b/lib/configuration.js#L187-L192).

Perhaps easier to explicitly state that it is *not* the same as the engine, and require a much simpler structure?

Reply:
Then maybe document that it must be an array, and each item in that array is either a plugin name (`string`, e.g., `remark-gfm`), or a tuple where the first item is the name, and the second an options object?

Reply:
As for externalizing, if possible, sure. But it’s all also complex because much of the code is about merging different config files, too.

Reply:
> Then maybe document that it must be an array

Currently an object is supported, just like `unified-engine`. As far as the JSON shape goes, this is compatible with `unified-engine`. Are you implying support for that should be removed? That would also mean the JSON schema needs to be redefined internally instead of referencing the schema store.

Reply:
It’s unified-engine-like, but it isn’t actually compatible.
Objects only make sense for strings (`{'remark-gfm': optionsForRemarkGfm}`), not for actually loaded plugins. MDX users likely know how to pass actually loaded plugins. So they’d probably not know about objects.

The JSON schema doesn’t match this implementation either, right?

I dunno, not blocking. Just think it’s better to start very strict, and add on to it later.

Reply:
> The JSON schema does match this implementation.

Then it doesn’t match the engine, because lists and presets are missing? 🤔

> I could imagine people might want to use other ways to configure the language server.

So do I, so it needs to be implemented at some point in the future.

> while it’s all still experimental

Sure.

> I think crashing on certain syntaxes (especially thinking of frontmatter) can be a deal breaker for many users.

Then show warnings? Currently the code exits silently on many things. to help users, you can show messages that an object was expected at L19, an array (or object) at L33-39, a plugin name at L51, etc